modules/ocl/cl_kernel.cpp

```diff
@@ -109,21 +124,99 @@ CLKernel::build_kernel (const XCamKernelInfo& info, const char* options)
+    bool load_cache = false;
+    struct timeval ts;
+
+    const char* cache_path = std::getenv ("XCAM_CL_KERNEL_CACHE_PATH");
+    if (NULL == cache_path) {
+        cache_path = _kernel_cache_path;
+    }
+
+    snprintf (
+        cache_filename, XCAM_MAX_STR_SIZE - 1,
+        "%s/%s",
+        cache_path, key_str);
+
     {
         SmartLock locker (_kernel_map_mutex);
+
         i_kernel = _kernel_map.find (key);
         if (i_kernel == _kernel_map.end ()) {
             SmartPtr<CLContext>  context = get_context ();
             single_kernel = new CLKernel (context, info.kernel_name);
             XCAM_ASSERT (single_kernel.ptr ());
-            ret = single_kernel->load_from_source (info.kernel_body, strlen (info.kernel_body), NULL, NULL, options);
-            XCAM_FAIL_RETURN (
-                ERROR, ret == XCAM_RETURN_NO_ERROR, ret,
-                "build kernel(%s) from source failed", key_str);
-            //kernel_map.insert (std::make_pair (key, single_kernel));
-            _kernel_map[key] = single_kernel;
-        } else
+
+            if (access (cache_path, F_OK) == -1) {
+                mkdir (cache_path, S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH);
+            }
+
+            ret = cache_file.open (cache_filename, "r");
+            if (ret == XCAM_RETURN_NO_ERROR) {
+                cache_file.get_file_size (read_cache_size);
+                if (read_cache_size > 0) {
+                    kernel_cache = (uint8_t*) xcam_malloc0 (sizeof (uint8_t) * (read_cache_size + 1));
+                    if (NULL != kernel_cache) {
+                        cache_file.read_file (kernel_cache, read_cache_size);
+                        cache_file.close ();
+
+                        ret = single_kernel->load_from_binary (kernel_cache, read_cache_size);
+                        xcam_free (kernel_cache);
+                        kernel_cache = NULL;
+
+                        XCAM_FAIL_RETURN (
+                            ERROR, ret == XCAM_RETURN_NO_ERROR, ret,
+                            "build kernel(%s) from binary failed", key_str);
+
+                        load_cache = true;
+                    }
+                }
+            } else {
+                XCAM_LOG_DEBUG ("open kernel cache file to read failed ret(%d)", ret);
+            }
+
+            if (load_cache == false) {
+                ret = single_kernel->load_from_source (info.kernel_body, strlen (info.kernel_body), &kernel_cache, &write_cache_size, options);
+                XCAM_FAIL_RETURN (
+                    ERROR, ret == XCAM_RETURN_NO_ERROR, ret,
+                    "build kernel(%s) from source failed", key_str);
+                //kernel_map.insert (std::make_pair (key, single_kernel));
+                _kernel_map[key] = single_kernel;
+            }
+        } else {
             single_kernel = i_kernel->second;
+        }
+    }
+
+    if (i_kernel == _kernel_map.end () &&
+            load_cache == false &&
+            NULL != kernel_cache) {
```

Comment:
1. i_kernel outside of lock is unsafe. 
2. this condition check is too complicated.
why not use a <bool need_write> for simple condition check?

Reply:
modified, please help review.